domain/resource/service/resource.go

```diff
@@ -267,14 +303,21 @@ func (s *Service) storeResource(
 		return errors.Errorf("getting resource: %w", err)
 	}
 
+	if args.Fingerprint.String() == res.Fingerprint.String() {
```

Comment:
There is a chance of a race here. The resource could be removed whilst adding a new one. This check should be done at the state layer.

Reply:
I don't believe having this check here is a problem. This method is inherently racy unless we put the blob in the store mid-transaction (which would be a terrible idea). The precondition that the fingerprints of the existing resource and the resource we are storing don't match could become untrue between us checking it and doing the Put.

If there are two stores trying to store the same blob one of the Puts will fail so only one Record will be run.

I'll add in a check to the state layer anyway to safeguard against this. If we are updating a resource with the same resource, we shouldn't return that the old one should be dropped as it is the same resource.

Reply:
As discussed in person, we can instead check the errors on the stores, if the fingerprint already exists in the store then they should return an error.